src/mon/AuthMonitor.cc

```diff
@@ -1941,6 +1898,69 @@ void AuthMonitor::_encode_key(const EntityName& entity,
   _encode_keyring(kr, entity, rdata, fmtr, wanted_caps);
 }
 
+/* Pass both, rdata as well as fmtr to enable printing of the key after
+ * update and set create to True to allow authorizing a new entity instead
+ * of updating its caps. */
+int AuthMonitor::_update_or_create_entity(const EntityName& entity,
+  const vector<string>& caps_vec, MonOpRequestRef op, stringstream& ds,
+  bufferlist* rdata, Formatter* fmtr, bool create_entity)
+{
+  stringstream ss;
+
+  if (!valid_caps(caps_vec, &ss)) {
+    return -EINVAL;
+  }
+
+  KeyServerData::Incremental auth_inc;
+  auth_inc.name = entity;
+  if (!create_entity &&
+      !mon.key_server.get_auth(auth_inc.name, auth_inc.auth)) {
+    ss << "couldn't find entry " << auth_inc.name;
+    return -ENOENT;
+  }
+
+  if (create_entity) {
+    auth_inc.auth.key.create(g_ceph_context, CEPH_CRYPTO_AES);
+  }
+
+  map<string, bufferlist> newcaps;
+  for (auto it = caps_vec.begin(); it != caps_vec.end(); it += 2)
+    encode(*(it+1), newcaps[*it]);
+  auth_inc.op = KeyServerData::AUTH_INC_ADD;
+  auth_inc.auth.caps = newcaps;
+
+  push_cephx_inc(auth_inc);
+
+  if (!create_entity) {
+    ss << "updated caps for " << auth_inc.name;
+  }
+
+  if (rdata != nullptr) {
+    _encode_auth(entity, auth_inc.auth, *rdata, fmtr, false, &newcaps);
+    rdata->append(ds);
+  }
+
+  string rs;
+  getline(ss, rs);
+  wait_for_finished_proposal(op, new Monitor::C_Command(mon, op, 0, rs,
+			     *rdata, get_last_committed() + 1));
+  return 0;
+}
+
+int AuthMonitor::_update_caps(const EntityName& entity,
+  const vector<string>& caps_vec, MonOpRequestRef op, stringstream& ds,
+  bufferlist* rdata, Formatter* fmtr) {
+  return _update_or_create_entity(entity, caps_vec, op, ds, rdata, fmtr,
+				  false);
+}
+
+int AuthMonitor::_create_entity(const EntityName& entity,
+  const vector<string>& caps_vec, MonOpRequestRef op, stringstream& ds,
+  bufferlist* rdata, Formatter* fmtr) {
```

Comment:
nit: let's stick to the local, already established convention for the opening brace (`{`).

Reply:
Sure.

I found that this local convention, of placing opening brace on a newline, brings more space and therefore makes code more readable, especially when a function has many parameters.

Reply:
Done.

Reply:
Done.